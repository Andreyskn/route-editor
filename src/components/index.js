export * from './PointsList';
export * from './Point';