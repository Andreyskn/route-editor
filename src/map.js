export class MapController {
	constructor(containerId) {
		if (!window.ymaps) throw new Error('Missing Map API script');

		this.map = null;
		this.points = {};
		this.polyline = null;
		this.order = [];

		window.ymaps.ready(() => this.init(containerId));
	}

	init = (containerId) => {
		try {
			this.map = new ymaps.Map(containerId, {
				center: [57.638724, 44.083593],
				zoom: 4,
				controls: [],
			});
		} catch {
			throw new Error(`Element with id="${containerId}" not found in DOM`);
		}
	}

	setActionCallback = (callback) => {
		window.ymaps.ready(() => this.map && this.map.events.add('actionend', callback));
	}

	addPlacemark = (id, name, order) => {
		this.order = order;

		const newPoint = new ymaps.Placemark(
			this.map.getCenter(),
			{ balloonContent: name },
			{ draggable: true, preset: 'islands#blueDotIcon' }
		);
		newPoint.events.add('dragend', this.drawLines);

		this.points[id] = newPoint;
		this.map.geoObjects.add(newPoint);

		this.drawLines();
	}

	removePlacemark = (id, order) => {
		this.order = order;

		this.map.geoObjects.remove(this.points[id]);
		delete this.points[id];

		this.drawLines();
	}

	changeOrder = (order) => {
		this.order = order;

		this.drawLines();
	}

	drawLines = () => {
		this.map.geoObjects.remove(this.polyline);

		this.polyline = new ymaps.Polyline(
			this.order.map(id => this.points[id].geometry.getCoordinates()),
			{},
			{ strokeWidth: 2, strokeColor: '#0000FF' }
		);

		this.map.geoObjects.add(this.polyline);
	}

	togglePlacemarkHighlight = (id, highlight) => {
		this.points[id].options.set('preset', highlight ? 'islands#greenDotIcon' : 'islands#blueDotIcon');
	}
}