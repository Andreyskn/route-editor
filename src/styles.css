*
{
	box-sizing: border-box;
}

body
{
	margin: 0;
}

.app
{
	display: flex;
	overflow: hidden;

	width: 100vw;
	height: 100vh;
}

.sidebar
{
	min-width: 230px;
	max-width: 380px;
	padding: 15px 0 0 15px;

	background-color: #466675;

	flex: 1 0 25%;
}

.input
{
	font-size: 16px;

	width: calc(100% - 15px);
	height: 40px;
	margin-bottom: 15px;
	padding: 0 10px;

	border: 1px solid black;
}

.points-list
{
	position: relative;

	overflow: auto;

	height: calc(100% - 55px);
	padding-right: 15px;
}

.point
{
	position: relative;

	display: flex;

	width: 100%;
	height: 50px;
	padding-right: 5px;

	user-select: none;

	border: 1px solid #b9e3f6;
	border-radius: 3px;
	background-color: aliceblue;

	align-items: center;
	justify-content: flex-end;
}

.point + .point:not(.clone)
{
	margin-top: 5px;
}

.dragged
{
	opacity: 0;
}

.clone
{
	position: absolute;
	top: 0;
	left: 0;

	width: calc(100% - 15px);

	pointer-events: none;

	opacity: .8;
	box-shadow: 1px 2px 5px 0 rgba(0,0,0,.75);
}

.point-overlay
{
	position: absolute;
	top: 0;
	left: 0;

	display: flex;

	width: 100%;
	height: 100%;
	padding-left: 10px;

	align-items: center;
}

.point-text
{
	font-size: 20px;

	display: block;
	overflow: hidden;

	max-width: calc(100% - 50px);

	white-space: nowrap;
	text-overflow: ellipsis;
}

.point-btn
{
	position: relative;
	z-index: 1;

	width: 40px;
	height: 40px;
	padding: 0;

	cursor: pointer;
	transition: all .25s;

	border: none;
	background-color: rgba(0, 0, 0, .04);

	flex-shrink: 0;
}

.point-btn:before,
.point-btn:after
{
	position: absolute;
	top: 50%;
	left: 50%;

	width: 2px;
	height: 15px;

	content: '';
	transition: all .25s;
	transform-origin: 0 0;

	background-color: #999;
}

.point-btn:before
{
	transform: rotate(45deg) translate(-50%, -50%);
}

.point-btn:after
{
	transform: rotate(-45deg) translate(-50%, -50%);
}

.point-btn:hover
{
	background-color: rgba(0, 0, 0, .1);
}

.point-btn:hover:before,
.point-btn:hover:after
{
	background-color: #000;
}

#map
{
	background-color: lightgray;

	flex: 1 0 75%;
}
