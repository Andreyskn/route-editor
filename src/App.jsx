import React, { Component } from 'react';

import { PointsList } from './components';
import MapController from './map';

const ENTER_KEY = 13;

export class App extends Component {

	_mapController = null;
	_input = null;

	state = {
		points: {},
		order: [],
	}

	componentDidMount() {
		this._mapController = new MapController('map');
		this._input && this._mapController.setActionCallback(this.setInputFocus);
	}

	setInputFocus = () => this._input.focus();

	generateId(ids) { return !ids.length ? 0 : Math.max(...ids) + 1 };

	addPoint = (e) => {
		const value = e.target.value;
		if (e.keyCode !== ENTER_KEY || value === '') return;

		const { points, order } = this.state;

		const newId = this.generateId(order);
		const updatedPoints = { ...points, [newId]: value };
		const updatedOrder = [...order, newId];

		e.target.value = '';

		this.setState({ points: updatedPoints, order: updatedOrder });
		this._mapController.addPlacemark(newId, value, updatedOrder);
	}

	removePoint = (id) => () => {
		const { points, order } = this.state;

		delete points[id];
		const updatedOrder = order.filter(i => i !== id);

		this.setState({ points, order: updatedOrder });
		this._mapController.removePlacemark(id, updatedOrder);
	}

	changeOrder = (order) => {
		this.setState({ order });
		this._mapController.changeOrder(order);
	}

	render() {
		const { points, order } = this.state;

		const inputProps = {
			type: 'text', className: 'input',
			placeholder: 'Новая точка маршрута',
			onKeyDown: this.addPoint, autoFocus: true,
		}

		const pointsListProps = {
			points, order,
			changeOrder: this.changeOrder,
			removePoint: this.removePoint,
			mapController: this._mapController,
		}

		return (
			<div className='app'>
				<div className='sidebar'>
					<input {...inputProps} ref={x => this._input = x} />
					<PointsList {...pointsListProps} />
				</div>
				<div id='map' />
			</div>
		)
	}
}