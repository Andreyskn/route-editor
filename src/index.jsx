import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import './styles.css';

import { PointsList } from './components';
import { MapController } from './map';

const ENTER_KEY = 13;

class App extends Component {

	_mapController = null;

	state = {
		points: {},
		order: [],
	}

	componentDidMount() {
		this._mapController = new MapController('map');
	}

	generateId = (ids) => !ids.length ? 0 : Math.max(...ids) + 1;

	addPoint = (e) => {
		const value = e.target.value;

		if (e.keyCode === ENTER_KEY && value !== '') {
			const { points, order } = this.state;

			const newId = this.generateId(order);
			const updatedPoints = { ...points, [newId]: value };
			const updatedOrder = [...order, newId];

			e.target.value = '';

			this.setState({ points: updatedPoints, order: updatedOrder });
		}
	}

	removePoint = (id) => () => {
		const { points, order } = this.state;

		delete points[id];
		const updatedOrder = order.filter(i => i !== id);

		this.setState({ points, order: updatedOrder });
	}

	changeOrder = (order) => {
		this.setState({ order });
	}

	render() {
		const { points, order } = this.state;

		return (
			<div className='app'>
				<div className='sidebar'>
					<input type='text' className='input' autoFocus placeholder='Новая точка маршрута' onKeyDown={this.addPoint} />
					<PointsList points={points} order={order} changeOrder={this.changeOrder} removePoint={this.removePoint} />
				</div>
				<div id='map' />
			</div>
		)
	}
}

ReactDOM.render(<App />, document.getElementById('root'));
