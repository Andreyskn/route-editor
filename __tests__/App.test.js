import React from 'react';
import { mount, shallow } from 'enzyme';
import { App } from '../src/App';
import MapController from '../src/map';

const mocks = {
	setActionCallback: jest.fn(),
	addPlacemark: jest.fn(),
	removePlacemark: jest.fn(),
	changeOrder: jest.fn(),
}

jest.mock('../src/map', () => jest.fn().mockImplementation(() => mocks));

describe('<App />', () => {

	const event = {
		target: { value: 'test' },
		keyCode: 13,
	}

	describe('Rendering', () => {

		it('renders correctly', () => {
			expect(mount(<App />)).toMatchSnapshot();
		})
	})

	describe('MapController calls', () => {

		beforeEach(() => {
			jest.restoreAllMocks();
		})

		it('on componentDidMount should create MapController instance and call setActionCallback', () => {
			mount(<App />);
			expect(MapController).toBeCalledWith(expect.any(String));
			expect(mocks.setActionCallback).toBeCalledWith(expect.any(Function));
		})

		it('should call addPlacemark method when new point is added', () => {
			const wrapper = mount(<App />);
			wrapper.instance().addPoint(event);
			expect(mocks.addPlacemark).toBeCalledWith(0, 'test', [0]);
		})

		it('should call removePlacemark method when point is removed', () => {
			const wrapper = mount(<App />);
			wrapper.instance().addPoint(event);
			wrapper.instance().removePoint(0)();
			expect(mocks.removePlacemark).toBeCalledWith(0, []);
		})

		it('should call changeOrder method when points order are reordered', () => {
			const wrapper = mount(<App />);
			wrapper.instance().changeOrder([1,2,3]);
			expect(mocks.changeOrder).toBeCalledWith([1,2,3]);
		})
	})

	describe('Component methods', () => {

		it('should update points on removePoint call', () => {
			const wrapper = shallow(<App/>);
			wrapper.setState({ points: { 0: 'a', 1: 'b' }, order: [0, 1] });
			wrapper.instance().removePoint(0)();
			const state = wrapper.state();
			expect(state.points).toEqual({ 1: 'b' });
			expect(state.order).toEqual([1]);
		})

		it('should generate unique id for new point', () => {
			const wrapper = shallow(<App/>);
			const generateIdSpy = jest.spyOn(wrapper.instance(), 'generateId');

			wrapper.instance().generateId([]);
			expect(generateIdSpy).lastReturnedWith(0);

			wrapper.instance().generateId([0,1,2]);
			expect(generateIdSpy).lastReturnedWith(3);
		})

		it('should focus input when setInputFocus is called', () => {
			const wrapper = shallow(<App/>);
			const mockFn = jest.fn();
			wrapper.instance()._input = { focus: mockFn };
			wrapper.instance().setInputFocus();

			expect(mockFn).toBeCalled();
		})
	})
});