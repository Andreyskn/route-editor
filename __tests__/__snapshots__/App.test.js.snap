// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Rendering renders correctly 1`] = `
<App>
  <div
    className="app"
  >
    <div
      className="sidebar"
    >
      <input
        autoFocus={true}
        className="input"
        onKeyDown={[Function]}
        placeholder="Новая точка маршрута"
        type="text"
      />
      <PointsList
        changeOrder={[Function]}
        mapController={null}
        order={Array []}
        points={Object {}}
        removePoint={[Function]}
      >
        <div
          className="points-list"
        >
          <div />
        </div>
      </PointsList>
    </div>
    <div
      id="map"
    />
  </div>
</App>
`;
